### ISSUE TITLE
WARNING: `pyenv init -` no longer sets PATH.

### ISSUE THREAD

**Issue by The-Judge** [pyenv/pyenv#1915]
Hi everyone,

I am aware of #1906 , but the solution seems to not do the trick for me:

I set everything to `.bashrc` and `.profile` as suggested in the output of `pyenv init`, but still there is this warning printed, causing issues with sftp and similar:

```
~ $ ssh judgepi /bin/true
WARNING: `pyenv init -` no longer sets PATH.
Run `pyenv init` to see the necessary changes to make to your configuration.
~ $
```

There are no other login/shell relevant files than `.bashrc` and `.profile`:
```
root@judgepi:~# ls -al | egrep -v ^d
total 152
-rw-------  1 root root 11209 May 10 20:02 .bash_history
-rw-r--r--  1 root root   873 May 10 19:52 .bashrc
-rw-r--r--  1 root root   748 Apr  3 09:48 .bashrc.backup
-rw-------  1 root root   102 Apr 22 11:29 .lesshst
-rw-------  1 root root   204 Nov 27 19:06 .netrc
-rw-r--r--  1 root root   252 May 10 19:40 .profile
-rw-------  1 root root    44 Nov 26 19:54 .python_history
-rw-r--r--  1 root root    66 Jul 28  2019 .selected_editor
-rw-------  1 root root 20183 May 10 20:00 .viminfo
lrwxrwxrwx  1 root root    30 Dec 12  2019 .virtualenvs -> /root/.local/share/virtualenvs
-rw-r--r--  1 root root   255 Apr 20 22:37 .wget-hsts
root@judgepi:~#
```

### Description
- Platform information (e.g. Ubuntu Linux 16.04): Raspbian GNU/Linux 10 (buster)
- OS architecture (e.g. amd64): armv7l
- pyenv version: 1.2.27-29-gfd3c891d
- Python version: 3.7.3
- C Compiler information (e.g. gcc 7.3): gcc 8.3.0
- [.bashrc and .profile gist](https://gist.github.com/The-Judge/94c9c599a980b08a65aa7ad9a9d956d6)

**Comment by aiguofer** [comment]
I was just trying to solve the same issue. The problem is that the warning is just a "dumb" message:

```
❯ pyenv init -
echo 'WARNING: `pyenv init -` no longer sets PATH.'
echo 'Run `pyenv init` to see the necessary changes to make to your configuration.'
export PYENV_SHELL=zsh
source '/home/aiguofer/.pyenv/libexec/../completions/pyenv.zsh'
command pyenv rehash 2>/dev/null
pyenv() {
  local command
  command="${1:-}"
  if [ "$#" -gt 0 ]; then
    shift
  fi

  case "$command" in
  activate|deactivate|rehash|shell)
    eval "$(pyenv "sh-$command" "$@")"
    ;;
  *)
    command pyenv "$command" "$@"
    ;;
  esac
}
```

It seems pretty difficult to ensure that the changes have been made, so it's hard to add the necessary checks in the script to disable the warning.

One solution could be to add a flag to disable the warning.. Maybe something like `eval $(pyenv init --disable-warning -)`

**Comment by albireox** [comment]
At least in `fish` shell this seems to have completely broken pyenv. The CLI is loaded but the shims don't seem to be working and it reverts to the system Python. Checking out tag 1.2.27 fixes this for now.

**Comment by aiguofer** [comment]
@albireox I believe that's because you did not follow the instructions (note that this shows the instructions for zsh since that's what I use, I imagine if you run `pyenv init` it should display the instructions for `fish`:

```
WARNING: `pyenv init -` no longer sets PATH.
Run `pyenv init` to see the necessary changes to make to your configuration.
❯ pyenv init
# Load pyenv automatically by appending
# the following to ~/.zshrc:

eval "$(pyenv init -)"

# And the following to ~/.profile:

eval "$(pyenv init --path)"

# Make sure to restart your entire logon session
# for changes to ~/.profile to take effect.
```

The crucial part here is that you need to set the `PATH` separately now to ensure shims work. 

This specific issue is about the `WARNING` message not going away after making the necessary changes.

**Comment by albireox** [comment]
This is the configuration I've been using for pyenv for years

```
set -gx PYENV_ROOT $HOME/.pyenv
set -gx fish_user_paths $PYENV_ROOT/bin $fish_user_paths
status --is-interactive; and pyenv init - --no-rehash| source
status --is-interactive; and pyenv virtualenv-init - | source
set -gx PYENV_VIRTUALENV_DISABLE_PROMPT 1
set -gx VIRTUAL_ENV_DISABLE_PROMPT 1
```

Nothing has changed on my side except that if I use the master branch now, in addition to the warning, the shims don't seem to work. 

I note the instructions for `fish` have changed and now it says to use

```
pyenv init --path | source
```

but if I do that then `pyenv-virtualenv` fails.

**Comment by aiguofer** [comment]
Switching to master branch IS a change. Master branch is where all new changes go, and if you update master branch all new changes are pulled in. There was a change in how `pyenv` is set up (see #1906 and #1649), and it requires changes on your side. That is why you are seeing a warning. The warning tells you exactly what you need to do:

```
WARNING: `pyenv init -` no longer sets PATH.
Run `pyenv init` to see the necessary changes to make to your configuration.
```

This means there was a change in the code. The original command to initialize `pyenv` no longer sets `PATH` (that is how shims get set up). By running `pyenv init` you can see the changes YOU need to make.

In my case, these are the important changes (I got this by running `pyenv init` as indicated by the warning message):

```
# Load pyenv automatically by appending
# the following to ~/.zshrc:

eval "$(pyenv init -)"
```

This initializes pyenv.

```
# And the following to ~/.profile:

eval "$(pyenv init --path)"
```

This sets up the `PATH`. 

```
# Make sure to restart your entire logon session
# for changes to ~/.profile to take effect.
```

This ensures everything is correctly loaded in your environment.

the TLDR:

BEFORE
```
eval "$(pyenv init -)"
```

NOW
```
eval "$(pyenv init -)" # in .zshrc
eval "$(pyenv init --path)" # in .profile
```

From your comment, my guess is you replaced one line with the other, but what you need to do is have both lines in their individual initialization files.

**Comment by native-api** [comment]
> Unless pyenv init --path is meant to be sourced only once

It _is_ meant to be sourced only once.

> Is it meant to output set -Ux PATH '/Users/albireo/.pyenv/shims' $PATH instead?

I've no experience with Fish so I just cargo-culted existing chunks.
You may wish to help with https://github.com/pyenv/pyenv/pull/1916 to fix Fish intructions.

**Comment by albireox** [comment]
Will do!

**Comment by andrewsiah** [comment]
Mine (mac os catalina) still fails after changing my ~/.profile to `export PYENV_ROOT="$HOME/.pyenv"
export PATH="$PYENV_ROOT/bin:$PATH"
eval "$(pyenv init --path)"`

and my ~/.zshrc to `export PATH="$HOME/.pyenv/bin:$PATH"
if command -v pyenv 1>/dev/null 2>&1; then
  eval "$(pyenv init -)"
fi`

**Comment by aiguofer** [comment]
> > The problem is that the warning is just a "dumb" message:
> 
> Not really... If you check `$PYENV_ROOT/libexec/pyenv-init`, it's added to the output only if `$PYENV_ROOT/shims` is not in `PATH`.

ahhhh interesting! I hadn't looked at the code. In my case, my issue was that I have a pretty convoluted `zsh` set up using prezto and antibody, while still trying to be portable and "modular" with various aliases, env configs, settings, etc across `zsh`, `bash` and my UI environment.. The only way I managed to do it was with some nasty sourcing in `.profile` that was causing `eval "$(pyenv init -)"` to be executed before `eval "$(pyenv init --path)"`. After just adding `export PATH="$HOME/.pyenv/shims:${PATH}"` to the top of `.profile` I no longer get the warning :) I'll have to play around with my config to find the best place to put it and still keep things nice and organized.

@andrewsiah did you log out and back in after making the changes?

**Comment by native-api** [comment]
I've made a PR with updated instructions for Bash and Zsh based on feedback: https://github.com/pyenv/pyenv/pull/1920.

Please review it and tell me:
* If the instructions (both README and `pyenv init`) are clear
* If they work for different distros
* If there's a better way

**Comment by phromo** [comment]
> I've made a PR with updated instructions for Bash and Zsh based on feedback: #1920.

I had a working env on Ubuntu 20.04 but couldn't make the warning go away. These instructions made my error clear; I had put the pyenv init at the end of .profile, but that doesn't work since .bashrc is sourced at the beginning.

**Comment by andrewsiah** [comment]
Does anyone have a fix? 

I've tried putting in all the commands in different permutations. 
<img width="644" alt="Screenshot 2021-05-12 at 6 39 43 PM" src="https://user-images.githubusercontent.com/31041391/117962301-a6f92200-b351-11eb-9732-706e143bae26.png">
(Here are my .profile, .zprofile, .zshrc)

Still getting this error when I do `pyenv init`
<img width="409" alt="Screenshot 2021-05-12 at 6 42 25 PM" src="https://user-images.githubusercontent.com/31041391/117962436-cc862b80-b351-11eb-9428-5e7873964ba0.png">

**Comment by JuPlutonic** [comment]
I didn't find here [above](https://github.com/pyenv/pyenv/issues/1915#issuecomment-839669813) the solution.
I thought that the change (the warnings) has more loose coupling with the place of your `pyenv init --path` (everywere, but before `pyenv init -`), but in fact that was not the case.
So the change is actually a bug, so I recommend everyone to check-out commit before the change (_debf1abd8bd624ba8f9f902f00e5c9098e99dede_).
```
cd ~/.pyenv; git reset --hard 783870759566a77d09b426e0305bc0993a522765
```
(_783870759566a77d09b426e0305bc0993a522765_ - good).

**Comment by native-api** [comment]
I've updated the instructions in #1920 based on feedback so far.

Please take a look and see if there's anything else that needs fixing/clarifying.

**Comment by native-api** [comment]
@aiguofer @andrewsiah @pgy866 @phromo @JuPlutonic 

Please try to follow the updated installation instructions in #1920 -- and tell me if there's still anything unclear in them.

If fact, if you have littered your shell configuration files from unsuccessful attempts -- you can first follow the _UNinstallation_ instructions (they are further down in the file, use the table of contents to find them).

**Comment by andrewsiah** [comment]
@native-api 

Thanks for the detailed instructions. 
I've done according to your suggestion (uninstall according to instructions, delete all pyenv code within .zshrc, .zprofile, .profile, to start afresh, and followed your installation instruction. 

Here are my new settings. 
<img width="641" alt="Screenshot 2021-05-13 at 12 10 20 PM" src="https://user-images.githubusercontent.com/31041391/118076212-33542500-b3e4-11eb-9a77-87db83e5f018.png">

Even after restarting, I'm still getting this warning,
<img width="636" alt="Screenshot 2021-05-13 at 12 10 35 PM" src="https://user-images.githubusercontent.com/31041391/118076248-46ff8b80-b3e4-11eb-9c8e-208f6d3e1933.png">

**Comment by native-api** [comment]
> Even after restarting, I'm still getting this warning,
> <img alt="Screenshot 2021-05-13 at 12 10 35 PM" width="636" src="https://user-images.githubusercontent.com/31041391/118076248-46ff8b80-b3e4-11eb-9c8e-208f6d3e1933.png">

You are not on the PR branch or even on `master`. Reinstall `pyenv` with `brew remove pyenv; brew install --head pyenv`.

I further cloned the pyenv doctor module and ran it. Which resulted in this.

What's your `brew --prefix` and `which -a pyenv`?

**Comment by JuPlutonic** [comment]
@native-api thx,
I tested it with [docker](https://gist.github.com/JuPlutonic/004879c3521a8cb246d9dcff5addedea) all seems Ok.
Main difference in my setup is what pyenv is installed from anyenv not from linuxbrew(Homebrew)).

**Comment by hr-curtissimo** [comment]
@native-api I uninstalled `pyenv` using `brew uninstall pyenv; brew install --head pyenv`. I added the lines specified from `pyenv init zsh`:

```
# In .zlogin
export PYENV_ROOT="$HOME/.pyenv"
export PATH="$PYENV_ROOT/bin:$PATH"
eval "$(pyenv init --path)"

# In .zshrc
eval "$(pyenv init -)"
```

When I start my shell, I get _two_ warning statements.

```
WARNING: `pyenv init -` no longer sets PATH.
Run `pyenv init` to see the necessary changes to make to your configuration.
WARNING: `pyenv init -` no longer sets PATH.
Run `pyenv init` to see the necessary changes to make to your configuration.
```

What am I doing wrong? 

I also run oh-my-zsh.

* **brew-prefix**: /usr/local
* **which -a pyenv**:  
  ```
  pyenv () {
	  local command
	  command="${1:-}" 
	  if [ "$#" -gt 0 ]
	  then
		  shift
	  fi
	  case "$command" in
		  (rehash | shell) eval "$(pyenv "sh-$command" "$@")" ;;
		  (*) command pyenv "$command" "$@" ;;
	  esac
  }
  ```

**Comment by davidjmeier** [comment]
I've been a long time pyenv user and haven't pulled the latest and greatest on my machine in a while.  This definitely tripped me up and for my working environments this is an unfortunate change - simply because I've tried to keep my dotfiles succint with respect to portability.  With that in mind I don't check in my `.profile` (ZSH user).  Now given how this works it appears that I have to add this to the mix.

Prior to this change initializing pyenv was very doable from `.zshrc` only.

**Comment by xo7799** [comment]
Hello, Because I don't understand the details, it may have something to do with the installation method. But If you are urgent to use an existing virtual environment, maybe you can use "source .pyenv/versions/**your_env**/bin/activate" in the terminal to enter the environment, and If you change the location, you need to find the environment location.

**Comment by xo7799** [comment]
> @native-api
> 
> 感謝您的詳細說明。
> 我按照你的建議做了（按照說明卸載，刪除.zshrc、.zprofile、.profile中的所有pyenv代碼，重新開始，然後按照你的安裝說明進行操作。
> 
> 這是我的新設置。
> <img alt="截圖 2021-05-13 at 12 10 20 PM" width="641" src="https://user-images.githubusercontent.com/31041391/118076212-33542500-b3e4-11eb-9a77-87db83e5f018.png">
> 
> 即使重新啟動後，我仍然收到此警告，
> <img alt="截圖 2021-05-13 at 12 10 35 PM" width="636" src="https://user-images.githubusercontent.com/31041391/118076248-46ff8b80-b3e4-11eb-9c8e-208f6d3e1933.png">
> 
> 我進一步克隆了 pyenv doctor 模塊並運行了它。這導致了這個。
> <img alt="截圖 2021-05-13 at 12 15 12 PM" width="637" src="https://user-images.githubusercontent.com/31041391/118076502-defd7500-b3e4-11eb-87ab-2332fcc14f0d.png">
> 
> 當然，當我之前檢查我的 brew 包時，它應該已經全部安裝在 brew 中。
> <img alt="截圖 2021-05-13 at 12 15 47 PM" width="469" src="https://user-images.githubusercontent.com/31041391/118076577-005e6100-b3e5-11eb-897d-e8a4d6a6c771.png">

If you have .bash_profile then you can try to modify it in the same way

**Comment by xo7799** [comment]
Maybe you can try the following command：
export PATH="~/.pyenv/bin:$PATH"
eval "$(pyenv init --path)"
eval "$(pyenv init -)"
# .bash_profile or need .zshrc
Depends on your terminal.

**Comment by xo7799** [comment]
Maybe you can try the following command：
export PATH="~/.pyenv/bin:$PATH"
eval "$(pyenv init --path)"
eval "$(pyenv init -)" 
.bash_profile or need .zshrc, depends on your terminal.

**Comment by execptionerror** [comment]
[Already solved WARNING: `pyenv init -` no longer sets PATH. Run `pyenv init` to see the necessary changes to make to your configuration.](https://exerror.com/warning-pyenv-init-no-longer-sets-path-run-pyenv-init-to-see-the-necessary-changes-to-make-to-your-configuration/)

**Comment by hwesterholt** [comment]
> I've been a long time pyenv user and haven't pulled the latest and greatest on my machine in a while. This definitely tripped me up and for my working environments this is an unfortunate change - simply because I've tried to keep my dotfiles succint with respect to portability. With that in mind I don't check in my `.profile` (ZSH user). Now given how this works it appears that I have to add this to the mix.
> 
> Prior to this change initializing pyenv was very doable from `.zshrc` only.

@davidjmeier Thank you for this - I have recently been forced to move from Linux (back) to macOS and this issue has been driving me nuts. I also keep my dotfiles synced across devices and have one little detail to add to the discussion: on Linux (openSUSE), I had kept the

```bash
export PYENV_ROOT="$HOME/.pyenv"
export PATH="$PYENV_ROOT/bin:$PATH"
eval "$(pyenv init --path)"
```
part in `.zprofile` (with a leading **z**!) and, after browsing the README, I had assumed that this is would work for `pyenv` (installed via git) on macOS as well. However, `pyenv init` told me to move these lines to `.profile` (without a **z**) instead; so I decided to disobey the README instructions which solved my issue and the warning has disappeared. Perhaps the README needs an update?

> Edit: After reading more into the issue I realized this was actually an Oh-My-ZSH issue. I simply needed to move my pyenv initialization above the OMZ initialization and everything went back to working as expected. Thanks all for such a great piece of FOSS!

As a fellow OMZ user, I cannot reproduce this issue. I initialize OMZ at the very top of my `.zshrc` whereas `eval "$(pyenv init -)"` is the last line.

**Comment by JuPlutonic** [comment]
So, if `echo $SHELL` gives zsh then prefer **.zprofile**, otherwise **.profile**.
